sdk/python/kfp/dsl/_pipeline_volume.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from kubernetes.client.models import (
+    V1Volume, V1PersistentVolumeClaimVolumeSource,
+    V1ObjectMeta, V1TypedLocalObjectReference
+)
+
+from . import _pipeline
+from ._pipeline_param import sanitize_k8s_name, match_serialized_pipelineparam
+from ._volume_snapshot_op import VolumeSnapshotOp
+
+
+class PipelineVolume(V1Volume):
```

Comment:
to be consistent with dsl.Container, could you name it dsl.Volume.

Reply:
I understand there is a `Container` class, but KFP does not expose it to the user, so we shouldn't be comparing with this. Instead, there is a `dsl.ContainerOp` and we are already consistent with it: We have a `ResourceOp`, a `VolumeOp` and a `VolumeSnapshotOp`, so everything is consistent.

On the other hand, a `dsl.PipelineVolume` is analogous to a `dsl.PipelineParam` and this *is* exposed to the user. It's an alternative way for passing information between steps, and carries *dependencies* that the compiler knows about, exactly as `PipelineParam` does.

So, we argue that the consistency is already there, and we should leave it as `PipelineVolume`, to really be consistent with what the user sees and uses.

Reply:
I see. I think the question is also discussed in another comment about the `after` method. To me, I am not fully convinced that it is special from other k8s resources. By leveraging PipelineParam, I can create other k8s resources from Op class. Maybe later I can add a SecretOp which returns a dsl.Secret object and I can inject it to ContainerOp(image_pull_secret = secret). 

I think we should expose dsl.Container and dsl.Volume to user. The reason is that they provides better programing experience than k8s native sdk. If user want to create a tfjob or pod resource, it's better that user can use similar interface as ContainerOp and PipelineVolume to build the nested containers and volumes. WDYT?

Reply:
@hongye-sun 
> Maybe later I can add a SecretOp which returns a dsl.Secret object and I can inject it to ContainerOp(image_pull_secret = secret).

This is a good example, we agree, and we have already covered this.
If `vop = VolumeOp(...)`, then you can always do `ContainerOp(volumes = {"/mnt": vop.volume})`, and you will have the implicit dependency. This works because `vop.volume` is a Volume and its name is a `PipelineParam`, so the ContainerOp will pick up the implicit dependency on `vop`.

Similarly, in your secret example, you only need to take a dependency on the *creation* of the secret,.

The problem is, referring to `vop.volume` only brings in dependencies from volume *creation* not volume use. A volume is a mutable resource, by definition, and it makes a difference whether you use it [mount it] before or after a certain `ContainerOp`.

But this distinction is lost if all `ContainerOp`s reference `vop` directly, and this is why we have added code in this PR so a `ContainerOp` can reference, e.g., `step1.volume` to get a new object, one that has extra dependencies on it, compared to `vop.volume`.

Note this part was a design goal of the original design doc #801, which received pretty positive reviews:
@StefanoFioravanzo https://github.com/kubeflow/pipelines/issues/801#issuecomment-463113410
> * Defining dependencies between volumes: Being able to declare step dependencies using the step_x.volume property is very straightforward and easy to understand for someone that has little to no knowledge of how Pipelines and K8S works.

@swiftdiaries https://github.com/kubeflow/pipelines/issues/801#issuecomment-463229853
>  I really like the dependency addition, it's very cool!

Finally:

> I think we should expose dsl.Container and dsl.Volume to user. The reason is that they provides better programing experience than k8s native sdk. If user want to create a tfjob or pod resource, it's better that user can use similar interface as ContainerOp and PipelineVolume to build the nested containers and volumes. WDYT?

I don't have an opinion on `dsl.Container`, at least not in the context of this PR, as I think it is orthogonal to getting this merged. I don't have an issue with renaming our `PipelineVolume` to `dsl.Volume` if you feel it's best for overall uniformity. And I agree that it is a much easier way to refer to existing PVCs by name [which can be in a `PipelineParam`], instead of having to use the K8s native API directly.

But I do believe that it makes sense to have a way to easily depend on the mutation produced on a volume by the execution of a specific `ContainerOp`, by referring to `step1.volume` instead of having to sprinkle `step2.after(step1)` calls manually throughout the code.

Reply:
@vkoukis, thanks for explaining the thoughts. I don't have strong opinion on that. I am fine to go with the change.